### DISCUSSION TITLE
Convert (homogeneous) data with edge type "to_heterogeneous"?

### DISCUSSION THREAD

**Discussion by isratnisa** [pyg-team/pytorch_geometric#3126] [Q&A] [answered]
Hi,
I want to use the new heterogeneous graph APIs to implement models like RGCN and use MUTAG and other common RGCN datasets.  Looks like these dataset are loaded as (homogeneous) Data object with a list of edge type, not as a HeteroData object where each edge types is stored in disjunct storage objects. You can convert heteroData to homogeneous data by using `to_homogeneous`, is the opposite direction also possible? We already have the edge type for each edge! I want to use `to_hetero` on my model which require the `metadata()` information from heteroData. Please let me know if there is a easy way around.

**Comment by wsad1** [comment, marked as answer]
For datasets like Mutag you could directly use [RGCNConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.RGCNConv), it has a different parameters for each `edge_type`.

But if you want to convert something like Mutag to HeteroData, one way to do it would be
```
from torch_geometric.data import HeteroData
data = HeteroData()
data['node_type1'].x = mutag.x # assuming there's just one node type.
for i in mutag.edge_type.unique():
    data['node_type1', str(int(i)), 'node_type1'] = mutag.edge_index[:,mutage.edge_type ==i]
```

**Reply by rusty1s** [reply]
Thanks @wsad1. The reverse conversion is indeed missing in the current package, and we should work on integrating it.

**Reply by wsad1** [reply]
@rusty1s  that's a good idea.
Would the reverse function be a classmethod of `HeteroData` with a signature that looks something like,
```
def to_hetero_data(cls, x:Tensor , node_type:List[str], edge_index:Tensor, edge_type:List[str],
                      edge_attr:Tensor)->HeteroData
    """
    node_type: List x.size(0) elements, indicates the type of each node.
    edge_type: List with edge_index.size(1) elements, indicates the type of each edge.
    """
```
I can create a PR for this if the above looks ok.

**Reply by isratnisa** [reply]
Thanks @wsad1. That did the job. 
What should be the `edge_type` though when you are using the RGCNConv? For example in the following demo code, when the model is called at the end ` out=model(data.edge_index_dict, ? )`, should the second param `edge_type` be a dict too? This param itself is a bit redundant on heteroData.

```
class RGCN_hetero(torch.nn.Module):
    def __init__(self):  
        super(RGCN_hetero, self).__init__()  
        self.conv1 = RGCNConv(data.num_nodes, 16, dataset.num_relations,  num_bases=30)  
        ...  
    def forward(self, edge_index, edge_type):  
        x = F.relu(self.conv1(None, edge_index, edge_type))  
        ...
model = RGCN_hetero()   
model = to_hetero(model, data.metadata(), aggr='sum').to(device)  

def train():  
  ...
  out = model(data.edge_index_dict, ? ) #  For homogeneous case it was a 1D array with edge type
```

**Reply by wsad1** [reply]
Since you are using to_hetero the message passing will be duplicated for each edge type as described [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html#automatically-converting-gnn-models). Which means each RGCN will now see a single `edge_type`, so `edge_type` can simply be set to `None`.

**Reply by rusty1s** [reply]
Note that combining `RGCNConv` and `to_hetero` does not make a lot of sense, since `RGCNConv` and `to_hetero` are similar in that they allow to operate on graphs with different relations. As an alternative, you can try the following:
```python
   class GNN(torch.nn.Module):
       def __init__(self, hidden_channels, out_channels):
           super().__init__()
           self.conv1 = SAGEConv((hidden_channels, hidden_channels), hidden_channels)
           self.conv2 = SAGEConv((hidden_channels, hidden_channels), out_channels)

       def forward(self, x, edge_index):
           x = self.conv1(x, edge_index).relu()
           x = self.conv2(x, edge_index)
           return x

model = GNN(hidden_channels=32, out_channels=16)
model = to_hetero(model, data.metadata(), aggr='sum')

emb = torch.nn.Embedding(num_nodes, 32)

x_dict = data.x_dict
x_dict['node_type'] = emb.weight
model(x_dict, data.edge_index_dict)
```

**Reply by isratnisa** [reply]
@rusty1s: I agree! Actually, the goal is to compare runtime between these two versions. Do you loose any performance because of using hetero data structure. We have similar implementation and observed difference in runtime. Thanks for the quick responses.

@wsad1 : Can't use None as edge_type without changing the nn module. Will try that. Thanks!

**Reply by rusty1s** [reply]
I see. There shouldn't be any major differences in runtime, but I'm interested to hear your findings. Please keep me updated :)

**Reply by rusty1s** [reply]
See https://github.com/pyg-team/pytorch_geometric/pull/3148

**Reply by isratnisa** [reply]
Ohh nice! Thank you!!

**Comment by mfaizan-ai** [comment]
@wsad1 how about when we have a multiple nodes type? I want to convert MUTAG datasets to a heterogenous graph structures with node and edge information like it is given in `heterData` object. I am not sure how to do that. Actually we want to convert TU datasets such MUTAG, COLLAB, IMDB-BINARY, etc etc to heterogenous graph for creating connection matrix.

**Reply by rusty1s** [reply]
This should work via `to_heterogeneous()` if `node_type` and `edge_type` are present attributes in your `Data` object. Can you confirm?
